noxfile.py

```diff
@@ -0,0 +1,173 @@
+    # TODO: uncomment if we want to use isort and docformatter
+    # session.run('isort', '--check', *PY_PATHS)
+    # session.run(
+    #     'docformatter',
+    #     '--check',
+    #     '--recursive',
+    #     '--wrap-summaries=100',
+    #     '--wrap-descriptions=100',
+    #     *PY_PATHS,
+    # )
+
+    # TODO: use flake8 instead of pyflakes
+    session.log('pyflakes b2')
+    output = subprocess.run('pyflakes b2', shell=True, check=False,
+                            stdout=subprocess.PIPE).stdout.decode().strip()
+    excludes = ['__init__.py']
+    output = [l for l in output.splitlines() if all(x not in l for x in excludes)]
+    if output:
+        print('\n'.join(output))
+        session.error('pyflakes has failed')
+    # session.run('flake8', *PY_PATHS)
+    session.run('pytest', 'test/static')
+
+
+@nox.session(python=PYTHON_VERSIONS)
+def unit(session):
+    """Run unit tests."""
+    install_myself(session)
+    session.install(*REQUIREMENTS_TEST)
+    session.run(
+        'pytest', '--cov=b2', '--cov-branch', '--cov-report=xml', '--doctest-modules',
+        *session.posargs, 'test/unit'
+    )
+    session.notify('cover')
+
+
+@nox.session(python=PYTHON_VERSIONS)
+def integration(session):
+    """Run integration tests."""
+    install_myself(session)
+    session.install(*REQUIREMENTS_TEST)
+    session.run('pytest', '-s', *session.posargs, 'test/integration')
+
+
+@nox.session(python=PYTHON_VERSIONS)
+def test(session):
+    """Run all tests."""
+    if session.python:
+        session.notify('unit-{}'.format(session.python))
+        session.notify('integration-{}'.format(session.python))
+    else:
+        session.notify('unit')
+        session.notify('integration')
+
+
+@nox.session
+def cover(session):
+    """Perform coverage analysis."""
+    session.install('coverage')
+    session.run('coverage', 'report', '--fail-under=75', '--show-missing')
+    session.run('coverage', 'erase')
+
+
+@nox.session(python=PYTHON_DEFAULT_VERSION)
+def build(session):
+    """Build the distribution."""
+    # TODO: consider using wheel as well
+    install_myself(session)
+    session.install(*REQUIREMENTS_BUILD)
+    session.run('liccheck', '-s', 'setup.cfg')
+    session.run('python', 'setup.py', 'check', '--metadata', '--strict')
+    session.run('rm', '-rf', 'build', 'dist', 'b2.egg-info', external=True)
+    session.run('python', 'setup.py', 'sdist', *session.posargs)
+
+
+@nox.session(python=PYTHON_DEFAULT_VERSION)
+def deploy(session):
+    """Deploy the distribution to the PyPi."""
+    session.install('twine')
+    session.run('twine', 'upload', 'dist/*')
+
+
+@nox.session(python=PYTHON_DEFAULT_VERSION)
+def doc(session):
+    """Build the documentation."""
+    install_myself(session)
+    session.install(*REQUIREMENTS_DOC)
+    session.cd('doc')
+    sphinx_args = ['-b', 'html', '-T', '-W', 'source', 'build/html']
+    session.run('rm', '-rf', 'build', external=True)
```

Comment:
no incremental doc building?

Reply:
We have incremental. From `CONTRIBUTING.md`:

## Documentation

To build the documentation and watch for changes (including the source code):

    nox -s doc

To just build the documentation:

    nox --non-interactive -s doc

Reply:
so... incremental builder works despite wiping the build/ directory?

Reply:
I mean `watch for changes` mode, not `incremental mode` in the sense that when you run the second time, it will rebuild only missing parts.
I prefer having only stateless sessions. I would go with stateful only if rebuilding from the scratch takes very long, but on my machine, it takes ~4s. For development, you can run it in interactive mode (default one!) and it will be watching for changes and build only differences.

Reply:
Right, I used to run it on HDD and it was not as fast. Also the old watcher sometimes missed something. Lets use nox native way then.